### PR TITLE
[BEAM-114] Add src distribution

### PR DESCRIPTION
**Pull request by jbonofre**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace "<Jira issue #>" in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by dhalperi** [comment]
R: @davorbonaci

File: src/main/descriptors/src.xml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  <formats>
+    <format>tar.gz</format>
+    <format>zip</format>
+  </formats>
+
+  <fileSets>
+    <fileSet>
+      <directory>${basedir}</directory>
+      <outputDirectory>/</outputDirectory>
+      <excludes>
```
Comment (davorbonaci) [inline src/main/descriptors/src.xml]: Could we add the whole contents of the `.gitignore` file here?

Also, I'd suggest to add a comment both here and in `.gitignore` that says that this list should include *everything* from .gitignore, but may include other things too.

File: src/main/descriptors/src.xml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (davorbonaci) [inline src/main/descriptors/src.xml]: This would be the only file in the top-level `src` directory, which would make that misleading to visitors.

I'd suggest to put the file, for the time being, inside `sdks/java` directory. The clear downside is that this is not really Java SDK-specific. However, we have `checkstyle` there right now, which has similar scope.

File: pom.xml [hunk 3]
```diff
@@ -143,6 +143,30 @@
...
+
+    <!-- src distribution -->
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-assembly-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>src</id>
+            <phase>package</phase>
```
Comment (davorbonaci) [inline pom.xml]: FYI: our build is getting pretty slow, and unnecessarily slowing developers down. The main cause is too much unnecessary plugins for that workflow. I think we should move all non-essential plugins to a profile.

That said, this is not isolated to this particular one. I'd keep as-is for now, but this is something we have to fix.

No action needed.

File: pom.xml [hunk 4]
```diff
@@ -143,6 +143,30 @@
   </profiles>
 
   <build>
+
+    <!-- src distribution -->
```
Comment (davorbonaci) [inline pom.xml]: FYI: if we are not doing a concurrent release of all components, we might need to create separate source packages on a per-component basis. I'd keep as-is for now, and cross that bridge later.

No action needed.

**Comment by davorbonaci** [comment]
Nice -- I didn't know this can be done so cleanly.

Left a few comments -- 2 FYI-only, 2 with minor action.

**Comment by jbonofre** [comment]
Rebase and update according to @davorbonaci comments.

File: sdks/java/src.xml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    <format>tar.gz</format>
+    <format>zip</format>
+  </formats>
+
+  <fileSets>
+    <fileSet>
+      <directory>${basedir}</directory>
+      <outputDirectory>/</outputDirectory>
+      <excludes>
+        <!-- NB: the exclude here should be sync with content of .gitignore -->
```
Comment (davorbonaci) [inline sdks/java/src.xml]: Add the comment to `.gitignore` too?

File: pom.xml [hunk 6]
```diff
@@ -344,6 +373,18 @@
           <artifactId>build-helper-maven-plugin</artifactId>
           <version>1.10</version>
         </plugin>
+
+        <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-release-plugin</artifactId>
```
Comment (davorbonaci) [inline pom.xml]: Unrelated change -- separate PR?

**Comment by davorbonaci** [comment]
LGTM
